src/backend/distributed/planner/multi_router_planner.c

```diff
@@ -377,19 +380,43 @@ RouterModifyTaskForShardInterval(Query *originalQuery, ShardInterval *shardInter
 	{
 		RelationRestriction *restriction = lfirst(restrictionCell);
 		List *originalBaserestrictInfo = restriction->relOptInfo->baserestrictinfo;
+		Var *relationPartitionKey = PartitionColumn(restriction->relationId,
+													restriction->index);
+		Param *uninstantiatedParameter = NULL;
 
 		/*
-		 * We haven't added the quals if all participating tables are reference
-		 * tables. Thus, now skip instantiating them.
+		 * We don't need to add restriction to reference tables given that they are
+		 * already single sharded and always prune to that single shard.
 		 */
-		if (allReferenceTables)
+		if (!relationPartitionKey)
 		{
-			break;
+			continue;
 		}
 
-		originalBaserestrictInfo =
-			(List *) InstantiatePartitionQual((Node *) originalBaserestrictInfo,
-											  shardInterval);
+		hashedOpExpressions = HashedShardIntervalOpExpressions(shardInterval);
+		Assert(list_length(hashedOpExpressions) == 2);
+
+		/*
+		 * Here we check whether the planner knows an equality between the partition column
+		 * and the uninstantiated parameter. If such an equality exists, we simply add the
+		 * shard restrictions.
+		 */
+		uninstantiatedParameter = UninstantiatedParameterForColumn(relationPartitionKey);
+		if (exprs_known_equal(restriction->plannerInfo, (Node *) relationPartitionKey,
```

Comment:
No, it should be ok.  There's some theoretical caveats, but otherwise...

Checking for relids would increase efficiency however - but we can postpone that to the point it's actually a measurable overhead...